protocols/upnp/src/behaviour.rs

```diff
@@ -396,27 +396,29 @@ where
         loop {
             match self.state {
                 GatewayState::Searching(ref mut fut) => match Pin::new(fut).poll(cx) {
-                    Poll::Ready(result) => match result {
-                        Ok(gateway) => {
-                            if !is_addr_global(gateway.external_addr) {
-                                self.state =
-                                    GatewayState::NonRoutableGateway(gateway.external_addr);
-                                log::debug!(
-                                    "the gateway is not routable, its address is {}",
-                                    gateway.external_addr
-                                );
-                                return Poll::Ready(ToSwarm::GenerateEvent(
-                                    Event::NonRoutableGateway,
-                                ));
+                    Poll::Ready(result) => {
+                        match result.expect("sender shouldn't have been dropped") {
```

Comment:
Are we sure this can't happen? We are already in a pattern match here so I'd suggest we just collapse these:

```rust
Poll::Ready(Ok(Ok(gateway))) => { }
```

Reply:
That makes sense to me. Are we missing something @thomaseizinger?

Reply:
> I can do as you suggested but the match bracket will be repeated code for the `Poll::Ready(Err(err))` bracket as `Canceled` is different than the `Box<dyn Error` returned by the inner `search_gateway`

IMO repeating patterns is cleaner than nesting them by YMMV :)

> how could the sender be dropped without first sending the response in our case? 

True! We can just panic in that branch then!